File: android/src/main/java/com/kakaotam/rnsdk/story/StoryApiClientModule.kt
package com.kakaotam.rnsdk.story

import com.facebook.react.bridge.*
import com.kakao.sdk.story.StoryApiClient
import com.kakao.sdk.story.model.LinkInfo
import com.kakao.sdk.story.model.Story
import com.kakao.sdk.story.model.StoryPostResult
import com.kakao.sdk.story.model.StoryProfile
import com.kakaotam.rnsdk.common.PromiseTransformer
import com.kakaotam.rnsdk.common.ReactToParam
import com.kakaotam.rnsdk.common.SdkInitializer
import com.kakaotam.rnsdk.story.model.*

class StoryApiClientModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "StoryApiClientModule"
    }

    @ReactMethod
    private fun delete(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, DeleteStroyParam::class.java)!!
        StoryApiClient.instance.delete(p.id) f@{ error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(true)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun isStoryUser(promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        StoryApiClient.instance.isStoryUser() f@{ isStoryUser: Boolean?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(isStoryUser!!)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun linkInfo(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, LinkInfoParam::class.java)!!
        StoryApiClient.instance.linkInfo(p.url) f@{ linkInfo: LinkInfo?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(linkInfo))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun postLink(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, PostLinkParam::class.java)!!
        StoryApiClient.instance.postLink(
            p.linkInfo,
            p.content,
            p.permission,
            p.enableShare,
            p.androidExecParam,
            p.iosExecParam,
            p.androidMarketParam,
            p.iosMarketParam
        ) f@{ storyPostResult: StoryPostResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(storyPostResult!!.id)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun postNote(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, PostNoteParam::class.java)!!
        StoryApiClient.instance.postNote(
            p.content,
            p.permission,
            p.enableShare,
            p.androidExecParam,
            p.iosExecParam,
            p.androidMarketParam,
            p.iosMarketParam
        ) f@{ storyPostResult: StoryPostResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(storyPostResult!!.id)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun postPhoto(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, PostPhotoParam::class.java)!!
        StoryApiClient.instance.postPhoto(
            p.imagePaths,
            p.content,
            p.permission,
            p.enableShare,
            p.androidExecParam,
            p.iosExecParam,
            p.androidMarketParam,
            p.iosMarketParam
        ) f@{ storyPostResult: StoryPostResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(storyPostResult!!.id)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun profile(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ProfileParam::class.java)!!
        StoryApiClient.instance.profile(
            p.secureResource
        ) f@{ profile: StoryProfile?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(profile))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun stories(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, StoriesParam::class.java)!!
        StoryApiClient.instance.stories(
            p.lastId
        ) f@{ stories: List<Story>?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(stories))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun story(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, StoryParam::class.java)!!
        StoryApiClient.instance.story(
            p.id
        ) f@{ story: Story?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(story))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun upload(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, StoryUploadParam::class.java)!!
        StoryApiClient.instance.upload(
            p.images
        ) f@{ uploadedPaths: List<String>?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(uploadedPaths))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }
}

File: ios/KakaoSdkStory/StoryApiClientModule.swift
//
//  StoryApiClientModule.swift
//  react-native-kakao-sdk-all
//
//  Created by kakao on 2022/05/23.
//

import Foundation

import KakaoSDKStory

@objc(StoryApiClientModule)
class StoryApiClientModule: NSObject {
    
    @objc(delete:withResolve:withReject:)
    func delete(_ param:NSDictionary,
                resolve:@escaping RCTPromiseResolveBlock,
                reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(DeleteStoryParam.self)!
            StoryApi.shared.delete(p.id) {
                (error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(true)
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(isStoryUser:withReject:)
    func isStoryUser(_ resolve:@escaping RCTPromiseResolveBlock,
                     reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            StoryApi.shared.isStoryUser() {
                (bool: Bool?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(bool ?? false)
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(linkInfo:withResolve:withReject:)
    func linkInfo(_ param:NSDictionary,
                  resolve:@escaping RCTPromiseResolveBlock,
                  reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(LinkInfoParam.self)!
            StoryApi.shared.linkInfo(url: p.url) {
                (linkInfo: LinkInfo?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(linkInfo?.toJsonString())
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(postLink:withResolve:withReject:)
    func postLink(_ param:NSDictionary,
                  resolve:@escaping RCTPromiseResolveBlock,
                  reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(PostLinkParam.self)!
            StoryApi.shared.postLink(content: p.content, linkInfo: p.linkInfo, permission: p.permission, enableShare: p.enableShare, androidExecParam: p.androidExecParam, iosExecParam: p.iosExecParam, androidMarketParam: p.androidMarketParam, iosMarketParam: p.iosMarketParam
            ) {
                (string: String?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(string)
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(postNote:withResolve:withReject:)
    func postNote(_ param:NSDictionary,
                  resolve:@escaping RCTPromiseResolveBlock,
                  reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(PostNoteParam.self)!
            StoryApi.shared.postNote(content: p.content,
                                     permission: p.permission,
                                     enableShare: p.enableShare,
                                     androidExecParam: p.androidExecParam,
                                     iosExecParam: p.iosExecParam,
                                     androidMarketParam: p.androidMarketParam,
                                     iosMarketParam: p.iosMarketParam) {
                (string: String?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(string)
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(postPhoto:withResolve:withReject:)
    func postPhoto(_ param:NSDictionary,
                   resolve:@escaping RCTPromiseResolveBlock,
                   reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(PostPhotoParam.self)!
            StoryApi.shared.postPhoto(content: p.content,
                                      imagePaths: p.imagePaths,
                                      permission: p.permission,
                                      enableShare: p.enableShare,
                                      androidExecParam: p.androidExecParam,
                                      iosExecParam: p.iosExecParam,
                                      androidMarketParam: p.androidMarketParam,
                                      iosMarketParam: p.iosMarketParam){
                (string: String?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(string);
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(profile:withResolve:withReject:)
    func profile(_ param:NSDictionary,
                 resolve:@escaping RCTPromiseResolveBlock,
                 reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ProfileParam.self)!
            StoryApi.shared.profile(secureResource: p.secureResource){
                (storyProfile: StoryProfile?, error: Error? ) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(storyProfile?.toJsonString());
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(stories:withResolve:withReject:)
    func stories(_ param:NSDictionary,
                 resolve:@escaping RCTPromiseResolveBlock,
                 reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(StoriesParam.self)!
            StoryApi.shared.stories(lastId: p.lastId){
                (stories: [Story]?, error: Error? ) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(stories?.toJsonString());
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(story:withResolve:withReject:)
    func story(_ param:NSDictionary,
               resolve:@escaping RCTPromiseResolveBlock,
               reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(StoryParam.self)!
            StoryApi.shared.story(id: p.id){
                (stroy: Story?, error: Error? ) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(stroy?.toJsonString());
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(upload:withResolve:withReject:)
    func upload(_ param:NSDictionary,
                resolve:@escaping RCTPromiseResolveBlock,
                reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(StoryUploadParam.self)!
            StoryApi.shared.upload(p.images) {
                (strings: [String]?, error: Error? ) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(strings?.toJsonString());
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
}
